cupy/cuda/device.pyx

```diff
@@ -15,7 +15,19 @@ from cupy import _util
 # It is always True as cuSOLVER library is always available in CUDA 8.0+.
 cusolver_enabled = True
 
-cdef object _thread_local = threading.local()
+
+class _ThreadLocalStack(threading.local):
+    def __init__(self):
+        self._device_stack = []
+
+    def push_device(self, value):
+        self._device_stack.append(value)
+
+    def pop_device(self):
+        return self._device_stack.pop()
```

Comment:
Could you follow the style used elsewhere in the codebase and write it as `cdef class _ThreadLocal:` without inheriting the `thread.local` class, and convert all methods to `cdef`? See, for example, 
https://github.com/cupy/cupy/blob/596d1af53b5793d3d52994c8b493ff42be453a8d/cupy_backends/cuda/stream.pyx#L13-L30

Reply:
Will do!

Reply:
There wasn't any state to encapsulate aside from the thread-local stack itself, so I routed it through an accessor function. That's not exactly the same as what you suggested, but it is consistent with other code in this file. LMK if the new version is okay.

Reply:
Hi @insertinterestingnamehere, I think you misunderstood me 😅 Could you take a look again at the `_ThreadLocal` class that I linked above? The state can be encoded as a cdef attribute (ex: `current_stream` there can be replaced by your `_device_stack`). So, your previous implementation was actually fine, just need to make it a cdef class, add a `cdef list _device_stack` attribute, etc.

Reply:
Oh, yah I totally misunderstood the suggestion. Thanks for clarifying! I'll switch it back on Monday.

Reply:
Actually Cython doesn't let us subclass from a non extension type with an extension type, so `cdef class` doesn't work here. Given that, what version do you prefer? I can switch it back to a plain Python class or just go with the accessor function as it currently is. Alternatively I can store the thread-local object as a member of an extension type and then move the accessor routine _get_handle to be an attribute on the new class instead. That's probably the best change, but it'll involve some additional refactoring outside of the code that handles the device stack. Really all three approaches work fine.

Reply:
Hi @insertinterestingnamehere, please kindly read my original https://github.com/cupy/cupy/pull/5083#discussion_r615008880 again. My suggestion was to *not* inherit the `threading.local` class (for the very same reason you correctly pointed out). The `_ThreadLocal` implementation that I linked to (which is adopted in many places in the codebase) should follow closely with your original implementation. 
>  Alternatively I can store the thread-local object as a member of an extension type and then move the accessor routine _get_handle to be an attribute on the new class instead.

This is indeed the idea.

Reply:
Okay, got it. For some reason I had assumed it'd be best to avoid changing the `_get_handle` code at all and was trying to interpret your suggestion in that light. Sure, I can refactor the `_get_handle` code here too.

Reply:
Wait, I am confused, why do we need to touch `_get_handle`?

Reply:
_get_handle currently routes all the thread-local data used in this file through _thread_local. By subclassing I let it continue doing that without any changes since _thread_local can still be used that way even with the added interface for device stack management. If I switch it to just being an attribute in the new class then I can't re-use the same thread-local object.

Any of the following would work: just using a different thread-local object, moving _get_handle to be a member function on the new class and using that interface instead of _get_handle (this one makes the most sense to me), or having _get_handle access the thread-local member of the new _thread_local object instead of using the _thread_local object itself. What do you prefer?

Reply:
The reason I tried writing _get_device_stack is because that roughly follows the idea behind _get_handle. It provides an accessor function to manage something in a module-level thread-local object.

Reply:
Right, I was being silly today, sorry...I agree making both `_get_handle` and `_get_device_stack` as the methods of the new `_ThreadLocal` class makes sense.

Reply:
No worries. I've pushed a new version that will hopefully work well. Thanks for your input!